book/src/pil/types.md

```diff
@@ -0,0 +1,249 @@
+references to columns are implicitly converted to expressions, as are integers and field elements.
+
+Example:
+```rust
+let x;
+let y;
+let f = || x + y;
+let g = || 7;
+f() == g();
+```
+The first two lines define the witness columns `x` and `y`.
+The next two lines define the utility functions `f` and `g`.
+The last line is at statement level and it is expected that it evaluates to a constraint, in this case, a polynomial identity.
+Because of that, `f()` is expected to evaluate to an expression, which means that the references
+to `x` and `y` that appear when evaluating `f` are already turned into expressions.
+The same is true about the `7` inside `g`.
+
+If `f` is called in a different context (for example inside a function that defines a constant column),
+then the column references `x` and `y` are interpreted to have the type `int -> fe` and thus
+adding them results in a type error.
+
+Since expressions are built from abstract column references, applying operators
+does not perform any operations but instead constructs an abstract expression structure / syntax tree.
+
+Expressions allow the following operators, which always construct new expressions:
+- `+`: additive combination of expressions
+- `-`: subtractive combination of expressions (also unary negation)
+- `*`: multiplicative combination of expressions
+- `**`: exponential combination of an expression with an integer constant
+- `'`: reference to the next row of a column, can only be applied directly to columns and only once
+
+The operator `==` on expressions constructs a constraint (see below);
+
+
+### Constraints
+
+Type name: `constr`
+
+Any evaluation of an expression at statement level needs to result in a constraint,
+or in an array of constraints.
+
+Generally, constraints include polynomial identities, plookups, permutations and connection identities.
+
+The only constraint currently constructible in the powdr-pil language are polynomial identities.
+These can be constructed from expressions by applying the `==` operator as in `x == 7`.
+
+Constraints do not allow any operators.
+
+
+## Implicit Conversions
+
+Usually, type conversions all have to be explicit using the built-in functions in the
+`std::convert` module, but there are some situations in which implicit type conversions happen:
+
+### Conversion to Expression
+
+If an expression is evaluated starting from statement level (i.e. the result is expected to be
+a constraint), references to columns are
+converted to expressions. If the column is a fixed column, it can still be called as
+a function with signature `int -> fe`.
+
+If an expression is combined with another value using a binary operator, the other value
+is converted to an expression according to the following rules:
+
+- field elements are converted without modification
+- non-negative integers less than the field modulus are converted to a field element and then to an expression
```

Comment:
Does that mean that a constraint like `-3 * a + b = 0` wouldn't work? I would have to write it as `b - 3 * a = 0` (because that would be parsed as subtraction with `3`) or replace `-3` with `modulus() - 3`?

Reply:
Yes, that is indeed a bit unfortunate. You can also write `0 - 3 * a + b = 0`.

Maybe we can relax the rules for conversion to expressions. I'll re-visit that in https://github.com/powdr-labs/powdr/pull/971 anyway.